app/models/iteration.rb

```diff
@@ -2,60 +2,25 @@ class Iteration < ActiveRecord::Base
   belongs_to :project
   has_many :tasks
 
-  # abstract the graph with the same iteration id
-  def abstract_graph
-    @tasks = Task.where(iteration: self.id)
-    graph = Hash.new
-    @tasks.each { |task| graph[task.id] = Taskedge.find_children task }
-    # JSON.generate(graph)
-    graph
-  end
-
-  # return the graph rank
-  def self.graph_rank graph
-    level = 0
-    graphlevel = Hash.new
-    root = Taskedge.find_root graph.keys
-    graphlevel[root] = level
-    children = graph[root]
-    visited = Array.new(children)
-    visited.append(root)
-    children.each{ |child| graphlevel[child] = level + 1 }
-    until children.empty?
-      newnode = children.shift
-      newchildren = graph[newnode]
-      newchildren.each { |child| graphlevel[child] = graphlevel[newnode] + 1 }
-      newchildren.delete_if { |child| visited.include? child }
-      visited.concat(newchildren)
-      children.concat(newchildren)
-    end
-    Iteration.level_up graphlevel,graph
-  end
-
   # calculate the percentage of accomplish of each task graph
-  def self.percentage_progress preTasks, devTasks, postTasks
-    precount = preTasks.select{ |task| task.task_status == 'finished'}.count*100.0
-    devcount = devTasks.select{ |task| task.task_status == 'finished'}.count*100.0
-    postcount = postTasks.select{ |task| task.task_status == 'finished'}.count*100.0
-    precountdanger = preTasks.select{ |task| task.task_status == 'danger'}.count*100.0
-    devcountdanger = devTasks.select{ |task| task.task_status == 'danger'}.count*100.0
-    postcountdanger = postTasks.select{ |task| task.task_status == 'danger'}.count*100.0
-    perdan = prepercent = devpercent = devdan = postpercent = postdan = 0
-    if preTasks.count > 0
-      prepercent = precount / preTasks.count
-      predan = precountdanger / preTasks.count
-    end
-
-    if devTasks.count > 0
-      devpercent = devcount / devTasks.count
-      devdan = devcountdanger / devTasks.count
-    end
+  def self.percentage_progress requestingTasks, planningTasks, executionTasks, deliveringTasks
```

Comment:
Similar to my previous comment, the calculations are very similar here.  Using a dictionary and using symbols can reduce the repetitiveness.